Port to matrix-rust-sdk-crypto

TobiasFella opened the pull request:
TODO:

- [ ] Test "all" possible verification scenarios
- [ ] Retry decrypting events once a new key has arrived
- [ ] Store new keys to backup
- [ ] Load keys from backup
- [ ] migration from current database to rust-sdk database (?)
- [ ] Clean up code, TODOs

TobiasFella committed:
Port to matrix-rust-sdk-crypto

KitsuneRal commented on .github/workflows/ci.yml:
```diff
@@ -172,20 +191,20 @@ jobs:
     - name: Configure libQuotient
       run: |
-        cmake -S $GITHUB_WORKSPACE -B $BUILD_PATH $CMAKE_ARGS -DQuotient_INSTALL_TESTS=ON
+        cmake -S $GITHUB_WORKSPACE -B build $CMAKE_ARGS -DQuotient_INSTALL_TESTS=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=1
```
Can you comment why you ended up with hardcoding `build` instead of using `$BUILD_PATH`?

TobiasFella replied:
Short version (also applies to the other comments about CI): I tried to make it work at all; had some problems IIRC mostly with environment variables. For reasons I don't fully remember things worked better with powershell.

My current plan is to make the CI work, then figure out which of the changes are actually necessary, then document those.

KitsuneRal commented on .github/workflows/ci.yml:
```diff
@@ -172,20 +191,20 @@ jobs:
+      run: cmake --build build --target update-api
 
     - name: Build and install libQuotient
+      shell: pwsh
```
Please document why :)

KitsuneRal commented on .github/workflows/ci.yml:
```diff
@@ -172,20 +191,20 @@ jobs:
-        ls ~/.local$BIN_DIR/quotest
+        # if [[ '${{ matrix.static-analysis }}' == 'sonar' ]]; then
+        #     BUILD_WRAPPER="${{ steps.sonar.outputs.build-wrapper-binary }} --out-dir build/sonar"
+        # fi
```
Anything wrong with sonar-scanner or did you just speed up going through CI this way?

TobiasFella replied:
I don't remember. I've re-enabled it for now, let's see what happens

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -89,6 +112,7 @@ Connection::~Connection()
 {
     qCDebug(MAIN) << "deconstructing connection object for" << userId();
     stopSync();
+
```
```suggestion
```

KitsuneRal commented on quotest/quotest.cpp:
```diff
@@ -224,20 +226,20 @@ TestManager::TestManager(int& argc, char** argv)
         clog << "Connected, server: " << c->homeserver().toDisplayString().toStdString() << '\n'
              << "Access token: " << c->accessToken().toStdString() << endl;
 
+        // We can't do this, as the crypto machine is already initialized and explodes when this is run
```
What this piece emulates is a client connecting to a homeserver, getting a token, immediately disconnecting, and connecting again with the obtained token. It would probably help to deconstruct the original connection before `newC->assumeIdentity()` is called? In that case the crypto machine in the original connection object will be dismantled before the one in the new connection is initialised.

KitsuneRal commented on quotest/quotest.cpp:
```diff
@@ -267,17 +269,21 @@ TestManager::TestManager(int& argc, char** argv)
 
 void TestManager::setupAndRun()
 {
+    qWarning() << 1;
```
Note to self: whole Quotest has to switch to `QDebug` framework, honestly. Originally it was a way to distinguish between the test output and the "normal" debug output but logging categories achieve the same goal in a much better way, and without weird desync between buffered `std::clog` and less-buffered `QDebug`.

KitsuneRal replied:
...and it's done in `dev` now.

KitsuneRal commented on quotest/quotest.cpp:
```diff
@@ -1005,7 +1011,6 @@ void TestManager::finalize(const QString& lastWords)
     c->logout().then(
         this, [this, lastWords] {
-            clog << lastWords.toStdString() << endl;
             QCoreApplication::exit(!testSuite ? -3
```
?

TobiasFella replied:
I think I removed this by accident when I was debugging something around these lines...

KitsuneRal commented on Quotient/connection.h:
```diff
@@ -352,53 +354,30 @@ class QUOTIENT_API Connection : public QObject {
+    // //! \brief Returns whether the device is known and supports end-to-end encryption.
+    // //!
+    // //! This might give unexpected results for users we're not tracking,
+    // //! i.e., users that we don't share an encrypted room with
```
```suggestion
    //! \brief Returns whether the device is known and supports end-to-end encryption.
    //!
    //! This might give unexpected results for users we're not tracking,
    //! i.e., users that we don't share an encrypted room with
```

KitsuneRal commented on Quotient/connection.h:
```diff
@@ -780,12 +759,25 @@ public Q_SLOTS:
 
+    QFuture<void> shareRoomKey(Room* room);
+    QString encryptRoomEvent(Room* room, const QByteArray& content, const QString& type);
+    QString decryptRoomEvent(Room* room, const QByteArray& event);
```
These should (ultimately) be in `Room`?

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -446,7 +543,48 @@ void Connection::sync(int timeout)
+            for (const auto& key : syncChanges->keys()) {
+                if (const auto& room = this->room(stringFromRust(key.room_id()))) {
+                    room->newSession(stringFromRust(key.session_id()));
+                }
```
```suggestion
                if (const auto& r = room(stringFromRust(key.room_id()))) {
                    r->newSession(stringFromRust(key.session_id()));
                }
```

KitsuneRal commented on Quotient/connection_p.h:
```diff
@@ -52,16 +54,19 @@ class Q_DECL_HIDDEN Quotient::Connection::Private {
     std::unordered_map<QString, EventPtr> accountData;
     QMetaObject::Connection syncLoopConnection {};
     int syncTimeout = -1;
+    std::optional<rust::Box<crypto::CryptoMachine>> cryptoMachine;
```
I think there's one extra level of indirection but maybe you already explored this path. `OlmMachine` is already under `Option<>`, I guess we could just add another value (`Uninitialised`?) to `enum InitError` and just store `cryptoMachine` as `crypto::CryptoMachine` here?

TobiasFella replied:
as far as i can tell, storing it as an object won't compile. Storing it as a raw pointer does compile fine, but crashes, as the object gets destroyed when we don't hold the `Box` somewhere.

TobiasFella replied:
I would like to get rid of the optional in some way, but I just haven't figured out what I'd need to do to still make it compile and work fine

KitsuneRal replied:
Let it stay like this for now then.

KitsuneRal commented on Quotient/events/filesourceinfo.cpp:
```diff
@@ -22,46 +32,42 @@ QByteArray Quotient::decryptFile(const QByteArray& ciphertext,
+}
+
+static QString stringFromRust(const rust::String& string) {
+    return QString::fromUtf8(bytesFromRust(string));
```
Do we really have to use `fromUtf8()` here? Key material is very much Latin-1, and that is unlikely to change.

KitsuneRal commented on Quotient/events/filesourceinfo.cpp:
```diff
@@ -22,46 +32,42 @@ QByteArray Quotient::decryptFile(const QByteArray& ciphertext,
-        qCWarning(E2EE) << "Decoded iv is too short for AES, need"
-                        << AesBlockSize << "bytes, got" << iv.size();
+
+    const auto result = crypto::decrypt_file(rust::Slice((std::uint8_t *) ciphertext.data(), ciphertext.size()), stringToRust(metadata.iv),  stringToRust(metadata.key.k), stringToRust(metadata.hashes["sha256"_L1]));
```
```suggestion
    const auto result = crypto::decrypt_file(rust::Slice(asWritableCBytes(ciphertext).data(), ciphertext.size()), stringToRust(metadata.iv),  stringToRust(metadata.key.k), stringToRust(metadata.hashes["sha256"_L1]));
```

KitsuneRal commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +35,107 @@ struct QUOTIENT_API EmojiEntry {
+        CREATED, //! The verification request has been newly created by us.
+        WAITINGFORREADY = CREATED, //! \deprecated Use CREATED instead
+        REQUESTED, //! The verification request was received from the other party.
+        INCOMING = REQUESTED, //! \deprecated Use REQUESTED instead
```
```suggestion
        CREATED, //! The verification request has been newly created by us
        DECL_DEPRECATED_ENUMERATOR(WAITINGFORREADY, CREATED),
        REQUESTED, //! The verification request was received from the other party
        DECL_DEPRECATED_ENUMERATOR(INCOMING, REQUESTED),
```

KitsuneRal commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +35,107 @@ struct QUOTIENT_API EmojiEntry {
+        READY, //! The verification request is ready to start a verification flow.
+        TRANSITIONED, //! The verification request has transitioned into a concrete verification flow. For example it transitioned into the emoji based SAS verification.
+        DONE, //! The verification flow that was started with this request has finished.
+        CANCELLED, //! The verification process has been cancelled.
```
```suggestion
        READY, //! The verification request is ready to start a verification flow
        TRANSITIONED, //! The verification request has transitioned into a concrete verification flow, e.g. into the emoji based SAS verification
        DONE, //! The verification flow that was started with this request has finished
        CANCELLED, //! The verification process has been cancelled
```

KitsuneRal commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +35,107 @@ struct QUOTIENT_API EmojiEntry {
+        KEYSEXCHANGED, //! The public keys have been exchanged and the short auth string can be presented to the user.
+        CONFIRMED, //! The verification process has been confirmed from our side, we’re waiting for the other side to confirm as well.
+        SASDONE, //! The verification process has been successfully concluded.
+        SASCANCELLED, //! The verification process has been cancelled.
```
```suggestion
        STARTED, //! The verification has been started, the protocols that should be used have been proposed and can be accepted
        ACCEPTED, //! The verification has been accepted and both sides agreed to a set of protocols that will be used for the verification process
        KEYSEXCHANGED, //! The public keys have been exchanged and the short auth string can be presented to the user
        CONFIRMED, //! The verification process has been confirmed from our side, we’re waiting for the other side to confirm as well
        SASDONE, //! The verification process has been successfully concluded
        SASCANCELLED, //! The verification process has been cancelled
```

KitsuneRal commented on Quotient/room.h:
```diff
@@ -806,6 +801,8 @@ class QUOTIENT_API Room : public QObject {
 
     QJsonArray exportMegolmSessions();
 
+    void newSession(const QString& newSession);
```
I'd rather use `newMegolmSession()`, so that there is less ambiguity with verification sessions. The same goes the other way around, e.g. `CryptoMachine` has `sessions` which actually are verification sessions, not Olm sessions.

KitsuneRal reviewed (commented):
Looked through most of the stuff, a few superficial comments and suggestions below. Overall LGTM so far.

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -326,31 +421,32 @@ void Connection::Private::completeSetup(const QString& mxId, bool newLogin,
         q->loadCapabilities();
         q->user()->load(); // Load the local user's profile
     }
+    auto doCompleteSetup = [this, mxId, deviceId, accessToken](){
```
```suggestion
    auto doCompleteSetup = [this, mxId, deviceId, accessToken](QKeychain::Job* = nullptr){
```
I'm afk and can't check it but that should make the lambda conform to `QFuture::then` interface below.

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -326,31 +421,32 @@ void Connection::Private::completeSetup(const QString& mxId, bool newLogin,
+        QDir(databasePath).removeRecursively();
+        runKeychainJob(new QKeychain::DeletePasswordJob(qAppName()), q->userId() % u"-Pickle"_s).then([doCompleteSetup](const auto &) {
+            doCompleteSetup();
+        });
```
```suggestion
        runKeychainJob(new QKeychain::DeletePasswordJob(qAppName()), q->userId() % u"-Pickle"_s).then(doCompleteSetup);
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -516,19 +653,78 @@ QJsonObject toJson(const DirectChatsMap& directChats)
+            auto deviceKeys = jsonFromRust(request.keys_upload_device_keys());
+            futures.append(q->callApi<UploadKeysJob>(deviceKeys.isEmpty() ? std::nullopt : std::make_optional(fromJson<DeviceKeys>(deviceKeys)), fromJson<OneTimeKeys>(jsonFromRust(request.keys_upload_one_time_keys())), fromJson<OneTimeKeys>(jsonFromRust(request.keys_upload_fallback_keys()))).then([id, this](const auto& job) {
+                (*cryptoMachine)->mark_keys_upload_as_sent(bytesToRust(job->rawData()), stringToRust(id));
+            }).QFuture<UploadKeysJob*>::then([](auto){}));
```
Something's off here. `JobHandle<T>::then` returns the same `JobHandle<T>` if the continuation returns `void`; and that should be convertible to `QFuture<void>` (but you might have to either convert explicitly or call `JobHandle::responseFuture()` to make the implicit conversion work). In other words, I made it specifically so that you don't have to jump through such hoops in this case. I'll only get to check it in a week when I'm back from vacation but it definitely shouldn't be this way.

TobiasFella replied:
with `JobHandle::responseFuture()` i can simplify the code somwhat; no idea how much sense it makes but it compiles fine and seems to work

KitsuneRal replied:
Thanks; and I'll separately check if making it implicitly convertible to `QFuture<void>`  can make any trouble otherwise.

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -516,19 +653,78 @@ QJsonObject toJson(const DirectChatsMap& directChats)
+    for (const auto &request : requests) {
+        auto id = stringFromRust(request.id());
+        auto type = request.request_type();
+        if (type == 0) { // Keys upload
```
Why not `switch()`?

TobiasFella replied:
no reason

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -554,8 +750,18 @@ void Connection::Private::consumeRoomData(SyncDataList&& roomDataList,
+                        if (cryptoMachine) {
+                            (*cryptoMachine)->update_tracked_users(userIds);
+                        }
+                    }
```
```suggestion
                    if (cryptoMachine && r->usesEncryption()) {
                        auto userIds = rangeTo<rust::Vec<rust::String>>(std::ranges::transform_view(r->memberIds(), &stringToRust));
                        (*cryptoMachine)->update_tracked_users(userIds);
                    }
```
Or something like this. And maybe `userIds` can even be `const`. And maybe we could do with `stringsToRust()` because it's not the only place this logic is needed.

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+                              const QString& targetDeviceId, const Event& event,
+                              bool encrypted)
+{
+}
```
Leftover?

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    }
+    auto request = requestResult->value();
+    callApi<SendToDeviceJob>(stringFromRust(request->to_device_event_type()), stringFromRust(request->to_device_txn_id()), fromJson<QHash<UserId, QHash<QString, QJsonObject>>>(jsonFromRust(request->to_device_messages())));
+    Q_EMIT newKeyVerificationSession(KeyVerificationSession::selfVerification(stringFromRust(request->verification_id()), this));
```
```suggestion
    emit newKeyVerificationSession(KeyVerificationSession::selfVerification(stringFromRust(request->verification_id()), this));
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+
+            q->callApi<SendToDeviceJob>(stringFromRust(request.event_type()), stringFromRust(id), fromJson<QHash<UserId, QHash<QString, QJsonObject>>>(jsonFromRust(request.messages()))).then([this, id](const auto& job) {
+                (*cryptoMachine)->mark_to_device_as_sent(bytesToRust(job->rawData()), id);
+            }, [=](const auto& job){});
```
This is the default value for the second parameter of `JobHandle::then`, you shouldn't need to spell it out, unless `=` does something non-trivial (doubt it).
```suggestion
            });
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+                (*cryptoMachine)->mark_to_device_as_sent(bytesToRust(job->rawData()), id);
+            }, [=](const auto& job){});
+        }
+    }, [=](const auto& job){});
```
Ditto
```suggestion
    });
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    } else {
+        auto json = jsonFromRust(outgoing->in_room_content());
+        auto transactionId = stringFromRust(outgoing->in_room_txn_id());
+        session->room()->postJson(u"m.key.verification.ready"_s, json);
```
```suggestion
        session->room()->postJson(KeyVerificationReadyEvent::TypeId, json);
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    } else {
+        auto json = jsonFromRust(startSas->in_room_content());
+        auto transactionId = stringFromRust(startSas->in_room_txn_id());
+        session->room()->postJson(u"m.key.verification.start"_s, json);
```
```suggestion
        session->room()->postJson(KeyVerificationStartEvent::TypeId, json);
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+        if (!session->room()) {
+            const auto& type = stringFromRust(request.to_device_event_type());
+            q->callApi<SendToDeviceJob>(type, stringFromRust(request.to_device_txn_id()), fromJson<QHash<UserId, QHash<QString, QJsonObject>>>(jsonFromRust(request.to_device_messages())));
+            if (type == u"m.key.verification.done"_s) {
```
```suggestion
            if (type == KeyVerificationDoneEvent::TypeId) {
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+            if (type == u"m.key.verification.done"_s) {
+                session->setState(KeyVerificationSession::DONE);
+            }
+            if (type == u"m.key.verification.cancel"_s) {
```
```suggestion
            if (type == KeyVerificationCancelEvent::TypeId) {
```

KitsuneRal replied:
```suggestion
            if (type == KeyVerificationCancelEvent::TypeId) {
```

KitsuneRal replied:
```suggestion
            if (type == KeyVerificationCancelEvent::TypeId) {
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+        } else {
+            const auto& type = stringFromRust(request.in_room_event_type());
+            session->room()->postJson(stringFromRust(request.in_room_event_type()), jsonFromRust(request.in_room_content()));
+            if (type == u"m.key.verification.done"_s) {
```
```suggestion
            if (type == KeyVerificationDoneEvent::TypeId) {
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+            if (type == u"m.key.verification.done"_s) {
+                session->setState(KeyVerificationSession::DONE);
+            }
+            if (type == u"m.key.verification.cancel"_s) {
```
```suggestion
            if (type == KeyVerificationCancelEvent::TypeId) {
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    } else {
+        auto json = jsonFromRust(outgoing->in_room_content());
+        auto transactionId = stringFromRust(outgoing->in_room_txn_id());
+        session->room()->postJson(u"m.key.verification.cancel"_s, json);
```
```suggestion
        session->room()->postJson(KeyVerificationCancelEvent::TypeId, json);
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+    if (!session->room()) {
+        q->callApi<SendToDeviceJob>(stringFromRust(request->to_device_event_type()), stringFromRust(request->to_device_txn_id()), fromJson<QHash<UserId, QHash<QString, QJsonObject>>>(jsonFromRust(request->to_device_messages())));
+    } else {
+        session->room()->postJson(u"m.key.verification.accept"_s, jsonFromRust(request->in_room_content()));
```
```suggestion
        session->room()->postJson(KeyVerificationAcceptEvent, jsonFromRust(request->in_room_content()));
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+Quotient::KeyVerificationSession* Connection::requestUserVerification(Room* room)
+{
+    auto session = KeyVerificationSession::requestUserVerification(room, this);
+    Q_EMIT newKeyVerificationSession(session);
```
```suggestion
    emit newKeyVerificationSession(session);
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+void Connection::receiveVerificationEvent(const QByteArray& fullJson)
+{
+    (*d->cryptoMachine)->receive_verification_event(bytesToRust(fullJson));
+    Q_EMIT verificationEventProcessed();
```
```suggestion
    emit verificationEventProcessed();
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+        return false;
+    }
+
+    return (*d->cryptoMachine)->is_e2ee_device(stringToRust(userId), stringToRust(deviceId));
```
```suggestion
    return d->cryptoMachine && (*d->cryptoMachine)->is_e2ee_device(stringToRust(userId), stringToRust(deviceId));
```

KitsuneRal commented on Quotient/connection.cpp:
```diff
@@ -1754,221 +1893,500 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
+        return false;
+    }
+
+    return (*d->cryptoMachine)->is_verified_device(stringToRust(userId), stringToRust(deviceId));
```
```suggestion
    return d->cryptoMachine && (*d->cryptoMachine)->is_verified_device(stringToRust(userId), stringToRust(deviceId));
```

TobiasFella commented:
- Updated rust dependencies
- Added migration for account and megolm sessions :partying_face: (I don't think we can migrate olm sessions; i also don't think we need to. I'll try to figure that out next week.)

TobiasFella commented:
- Turns out we do need to migrate the olm sessions; i also figured out how to do so. It's already implemented cleaner than the megolm migration, but also still needs error handling
- Next: migrating CS keys (and backup version, i think)
- Figure out whether migrating all olm and megolm sessions is a good idea, since i guess there might be many of them

redstrate commented on .github/workflows/ci.yml:
```diff
@@ -124,6 +124,25 @@ jobs:
+      run: |
+        curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs > rustup-init.sh
+        sh rustup-init.sh -y --default-toolchain none
+        rustup target add "x86_64-apple-darwin"
```
Do you also want to run `rustup --version` as a sanity check too, or am I misunderstanding why you did it in Windows?

TobiasFella replied:
added. It's was probably just for debugging, but can't hurt

redstrate commented on .github/workflows/ci.yml:
```diff
@@ -135,21 +154,29 @@ jobs:
+      if: startsWith(matrix.os, 'macos')
+      run: |
+        cd ..
+        git clone -b v0.13.0 https://github.com/frankosterfeld/qtkeychain.git
```
That seems like a pretty old version of qtkeychain :stuck_out_tongue:

TobiasFella replied:
Yes, I did this to fix problems because new versions weren't working for reasons i can't properly debug without the right hardware

redstrate commented on .github/workflows/ci.yml:
```diff
@@ -135,21 +154,29 @@ jobs:
-        git clone https://gitlab.matrix.org/matrix-org/olm.git
-        cmake -S olm -B olm/build $CMAKE_ARGS
-        cmake --build olm/build --target install
+        git clone https://github.com/corrosion-rs/corrosion
```
Might be smart to pin this on a version?

redstrate commented on CMakeLists.txt:
```diff
@@ -90,31 +90,31 @@ set(${PROJECT_NAME}_INSTALL_INCLUDEDIR
 
-find_package(${Qt} ${QtMinVersion} REQUIRED Core Network Gui Test Sql)
+find_package(${Qt} ${QtMinVersion} REQUIRED COMPONENTS Core Network Gui Test Sql)
+find_package(${Qt} ${QtMinVersion} COMPONENTS CorePrivate)
```
Isn't this already done a few lines down?

redstrate commented on Quotient/connection.cpp:
```diff
@@ -306,15 +313,163 @@ void Connection::Private::loginToServer(LoginArgTs&&... loginArgs)
+        .unwrap()
+        .then([](QFuture<Job*> writeFuture) {
+            if (const Job* const writeJob = writeFuture.result();
+                writeJob && writeJob->error() != Error::NoError) //
```
```suggestion
                writeJob && writeJob->error() != Error::NoError)
```

redstrate commented on Quotient/connection.cpp:
```diff
@@ -306,15 +313,163 @@ void Connection::Private::loginToServer(LoginArgTs&&... loginArgs)
+    bool hasDb = QFileInfo(QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) + u'/' + mxIdForDb + "/quotient_%1.db3"_L1.arg(q->deviceId())).exists();
+    if (hasDb) {
+        auto db = QSqlDatabase::addDatabase(u"QSQLITE"_s, "Quotient_"_L1 + q->deviceId());
+        const QString databasePath{ QStandardPaths::writableLocation(QStandardPaths::AppDataLocation) % u'/' % mxIdForDb };
```
`databasePath` again? (and further decls)

redstrate commented on Quotient/connection.cpp:
```diff
@@ -306,15 +313,163 @@ void Connection::Private::loginToServer(LoginArgTs&&... loginArgs)
+            sessions.push_back(OlmSessionData {
+                .libolm_pickle = stringToRust(query.value(u"pickle"_s).toString()),
+                .sender_key = stringToRust(query.value(u"senderKey"_s).toString()),
+                .creation_time = std::uint64_t(query.value(u"lastReceived"_s).toDateTime().toSecsSinceEpoch()), // This is a lie
```
Why is it a lie? Sounds like it needs to be a TODO

TobiasFella replied:
The rust sdk wants the creation time, i don't really know why. We don't have that stored, so we can't pass the correct data. I don't see any way we can change that, so no point in a TODO

redstrate commented on Quotient/connection.cpp:
```diff
@@ -306,15 +313,163 @@ void Connection::Private::loginToServer(LoginArgTs&&... loginArgs)
+                    auto cipher = QByteArray::fromBase64(query.value(u"cipher"_s).toString().toLatin1());
+                    auto iv = QByteArray::fromBase64(query.value(u"iv"_s).toString().toLatin1());
+                    if (iv.size() < 16) {
+                        qCWarning(E2EE) << "Corrupt iv at the database record for" << name;
```
I don't know when this failure occurs, but we may also want to print `iv` - or if that's some kind of secure data then at least the size to help debugging.

TobiasFella replied:
I'm not sure why we're even checking this explicitely. Sure it can happen, but so can a bunch of other things. I don't there's a point in printing it, it would just look like random bytes and if this is broken (and not just the database corrupt), there would be something so fundamentally broken that we would have already found it

redstrate commented on Quotient/connection.cpp:
```diff
@@ -306,15 +313,163 @@ void Connection::Private::loginToServer(LoginArgTs&&... loginArgs)
+                    query.bindValue(u":name"_s, name);
+                    query.exec();
+                    if (!query.next()) {
+                        return QByteArray();
```
Does this failure have to be reported or is it expected?

TobiasFella replied:
These failures are very expected (if the neochat session is not yet cross-signed). This was missing the required handling on the rust side; I added it now

redstrate commented on Quotient/connection.cpp:
```diff
@@ -499,34 +647,183 @@ void Connection::syncLoopIteration()
+
+void Connection::Private::processOutgoingRequests()
+{
+    if (!cryptoMachine) {
```
Would this also be useful to have an assert for? I would think we always expect a `cryptoMachine`

TobiasFella replied:
In theory yes, but in practice the startup sequence is a bit messy. Currently, this is actually hit and fully preventing that would require more re-engineering than i want to do at the moment.

redstrate commented on Quotient/connection.cpp:
```diff
@@ -499,34 +647,183 @@ void Connection::syncLoopIteration()
+        const auto id = stringFromRust(request.id());
+        const auto type = request.request_type();
+        switch (type) {
+            case 0: { // Keys upload
```
Is there really no enum exposed? :(

TobiasFella replied:
This was from before i knew you could bridge enums. Fixed

redstrate commented on Quotient/connection.cpp:
```diff
@@ -1768,249 +2007,545 @@ QJsonObject Connection::decryptNotification(const QJsonObject& notification)
-    return d->encryptionData->deviceKeys.value(userId).keys();
+    auto requestResult = (*d->cryptoMachine)->request_self_verification();
+    if (requestResult->has_error()) {
+        return;
```
Do we want to report that error?

TobiasFella replied:
already done on the rust side

redstrate commented on Quotient/room.cpp:
```diff
@@ -3022,40 +2859,28 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
-                    auto session = new KeyVerificationSession(evt, q);
+                    auto json = evt->fullJson();
+                    if (!json.contains(u"room_id"_s)) {
+                        json[u"room_id"_s] = q->id(); // Synapse seems to send broken events?
```
That seems like quite the comment, does that need to be followed up on?

TobiasFella replied:
ehm. Yes. Let me add a TODO so i don't miss it...

TobiasFella replied:
Turns out, that's just how matrix works. The events we get from sync don't have a room_id field; for additional confusion, we *do* have it for decrypted events...

redstrate commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +37,111 @@ struct QUOTIENT_API EmojiEntry {
+        CONFIRMED, //! The verification process has been confirmed from our side, we’re waiting for the other side to confirm as well
+        SASDONE, //! The verification process has been successfully concluded
+        SASCANCELLED, //! The verification process has been cancelled
+        SASNOTFOUND,
```
Missing doc comment for this variant

redstrate commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +37,111 @@ struct QUOTIENT_API EmojiEntry {
+        DONE, //! The verification flow that was started with this request has finished
+        CANCELLED, //! The verification process has been cancelled
+        CANCELED = CANCELLED,
+        NOTFOUND,
```
Also missing a doc comment for this variant

redstrate commented on Quotient/keyverificationsession.cpp:
```diff
@@ -3,745 +3,218 @@
+    if (m_state != READY) {
+        return;
     }
+    //TODO: resolve glare
```
Glare?

TobiasFella replied:
Not relevant; removed.

redstrate commented on Quotient/crypto-sdk/Cargo.toml:
```diff
@@ -0,0 +1,49 @@
+
+
+[target.'cfg(macos)'.dependencies]
+cxx = { version = "=1.0.138", default-features = true, features = [ "c++20" ] }
```
```suggestion
cxx = { version = "1", default-features = true, features = [ "c++20" ] }
```

TobiasFella replied:
nope :) bizarre compilation errors in cxx

redstrate commented on Quotient/crypto-sdk/src/cryptomachine.rs:
```diff
@@ -0,0 +1,2078 @@
+use matrix_sdk_common::{
```
Missing REUSE?

redstrate commented on Quotient/crypto-sdk/src/cryptomachine.rs:
```diff
@@ -0,0 +1,2078 @@
+        Box::new(match result {
+            Ok(requests) => ToDeviceRequestResult::Ok(requests),
+            Err(error) => {
+                eprintln!("Failed to share room key: {:?}", error);
```
I picked this random one, but are we okay with `eprintln` for errors? This will pop up unformatted in the log.

TobiasFella replied:
Switched to log::error. Or did you have something else in mind?

TobiasFella replied:
for these ones, we could also move them to the C++ side i guess

redstrate commented on Quotient/crypto-sdk/src/verification.rs:
```diff
@@ -0,0 +1,48 @@
+        if let OutgoingVerificationRequest::ToDevice(request) = &self.1 {
+            request.event_type.to_string()
+        } else {
+            panic!("This is not a to-device event")
```
Do you actually want to panic! here? :D

TobiasFella replied:
Debatable. If this is hit, something is *really* wrong in the code, so conceptually it's an assert. The nice thing about using panic is that it has the `Never` type, so I don't have to produce an actual (invalid) value of the respective request types...

redstrate reviewed (commented):
Initial review drop, will do more when I'm feeling better

redstrate commented on Quotient/crypto-sdk/src/cryptomachine.rs:
```diff
@@ -0,0 +1,2104 @@
+
+impl OutgoingRequestResult {
+    pub(crate) fn has_error(&self) -> bool {
+        matches!(self, OutgoingRequestResult::Err(_))
```
Stylistic nitpick, but if you didn't know you should be able to write it like this: `Self::Err(_)` and it would resolve to `OutgoingRequestResult` here.

(If you prefer the more explicit version, just ignore the comment)

TobiasFella replied:
done

redstrate commented on Quotient/crypto-sdk/src/cryptomachine.rs:
```diff
@@ -0,0 +1,2104 @@
+impl BackupRequest {
+    pub(crate) fn transaction_id(&self) -> String {
+        self.transaction_id
+            .clone()
```
nit: Do these functions (and ditto) need to use `clone`? Clippy would tell you I guess.

redstrate commented on Quotient/keyverificationsession.h:
```diff
@@ -34,153 +37,111 @@ struct QUOTIENT_API EmojiEntry {
+        ACCEPTED, //! The verification has been accepted and both sides agreed to a set of protocols that will be used for the verification process
+        KEYSEXCHANGED, //! The public keys have been exchanged and the short auth string can be presented to the user
+        CONFIRMED, //! The verification process has been confirmed from our side, we’re waiting for the other side to confirm as well
+        SASDONE, //! The verification process has been successfully concluded
+        SASCANCELLED, //! The verification process has been cancelled
+        SASNOTFOUND, //! The verification session is not found / not transitioned to SAS
```
Why are some of these variants prefixed with SAS? I guess it's not to conflict with the ones in `State`, but then everything should be prefixed really(?)

TobiasFella replied:
yup

redstrate commented on Quotient/keyverificationsession.cpp:
```diff
@@ -3,745 +3,203 @@
-}
-
-void KeyVerificationSession::trustKeys()
+QVector<EmojiEntry> KeyVerificationSession::sasEmojis()
```
Eh, `QVector`? If we want to use a QList, let's just call it a `QList`.

TobiasFella replied:
yup

redstrate commented:
Do we want https://github.com/quotient-im/libQuotient/pull/863 in before this or can they be merged unrelated to each other?

TobiasFella commented:
> Do we want #863 in before this or can they be merged unrelated to each other?

I guess that doesn't matter too much, whichever goes in later must be adapted

nvrWhere commented:
So currently when I try to verify my session with NeoChat it all starts fine but no emojis appear once both sides have started

nvrWhere commented:
I've been using this for months and haven't noticed which in this case is great. It's worlds better than where we are now. I think we ship this.

KitsuneRal reviewed (commented):
Functionally, this looks very reasonable overall by now. Thanks for the huge work done! The only thing I discovered is actually in the build configuration. There's something off with the path to `lib.rs.h` because my initial attempt to build the PR failed because the compiler couldn't find the header (despite it having been generated obviously). Also, the fact that it relies on a very new Corrosion version that has to be manually installed is a bit unfortunate; but I really don't know if we should wait for it to land in distro repos or just commit the generated files instead... which might solve the `lib.rs.h` path problem along the way, too?

TobiasFella commented:
Ugh, thought the header generation thing was fixed. Weird that it works fine on the CI, then... I'll have a look tomorrow.

WRT corrosion, i think we should just ship it like it is now. The best way to make a distro update a dependency is to just enforce it.

TobiasFella commented:
I really don't know what's up with lib.rs.h... It works completely fine here. Is there anything you can say about when it is (and isn't) failing for you?

KitsuneRal commented:
Right; I really should've given more than just a "doesn't build here" note. Buut, now that I tried to build it from clean slate, one more thing surfaced:
```
   Compiling matrix-sdk-sqlite v0.16.0
error[E0308]: mismatched types
    --> src/cryptomachine.rs:1887:37
     |
1887 | ...                   pickle: session.pickle(),
     |                               ^^^^^^^^^^^^^^^^ expected `SessionPickle`, found `vodozemac::olm::SessionPickle`
     |
note: there are multiple different versions of crate `vodozemac` in the dependency graph
    --> /var/home/axy/dev/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/vodozemac-0.9.0/src/olm/session/mod.rs:481:1
     |
 481 | pub struct SessionPickle {
     | ^^^^^^^^^^^^^^^^^^^^^^^^ this is the expected type
     |
    ::: /var/home/axy/dev/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/vodozemac-0.10.0/src/olm/session/mod.rs:510:1
     |
 510 | pub struct SessionPickle {
     | ------------------------ this is the found type
     = help: you can use `cargo tree` to explore your dependency tree
```
`cargo tree` shows that indeed, we get the older vodozemac with `matrix-sdk` crates but then also get vodozemac-0.10 via our own Cargo.toml. That was what I could see from absent `~/.cargo`, no `Cargo.lock`, and with a freshly done CMake configuration. If I manually remove the new vodozemac from `Cargo.lock` then the build goes through (even without that former error about `lib.rs.h` I've got, which is also peculiar). In any case, that definitely needs fixing: we should only pull the version of vodozemac that other crates are using, not just anything with `0` major version. I'm not sure if pinning down vodozemac 0.9 in `Cargo.toml` is the best way but that's as good as I can come up with for now.

redstrate commented:
Without testing, IIRC you can specify a "*" and Cargo should be smart enough to pick the version used by the other crates.

TobiasFella commented:
Fixed by not depending on vodozemac directly.

KitsuneRal commented:
Hm, that seemed to work although when I tried the same locally before it was complaining about not finding the crate. Whatever works.
I'll give it a couple more runs but (aside from panics in case of some errors) this does seem good to go.

KitsuneRal commented:
I don't want to hold it off for another month while I'm away; so I guess that's it.
